README.md
# EPU_GROUP_AFIS

Takes the XML metadata from an EPU session that used AFIS (_Aberration Free
Image Shift_) for faster acquisition and organizes the collected micrograph
movies into RELION Optics Groups. In RELION version 3.1 where these groups were
introduced they can be used to further reduce the effects of high-order
aberrations on the data induced by collecting using large amounts of beam-image
shift.

There is a Notebook here that goes through how the processing is done and then a
script that is for actual processing.

## Requirements

It should be consistent across older versions of Python, but this was what I
tested it using:

 * Python (_Version >= 3.7.6_)
 * numpy (_Version >= 1.18.1_)
 * scikit-learn (_Version >= 0.22.1_) 
 * matplotlib (_Version >= 3.1.3_)

## Usage
```
usage: EPU_Group_AFIS.py [-h] --xml_dir XML_DIR [--n_clusters N_CLUSTERS]
                         [--apix APIX] [--mtf_fn MTF_FN] [--voltage VOLTAGE]
                         [--cs CS] [--q0 Q0] [--ftype FTYPE]
                         [--movie_dir MOVIE_DIR] [--output_fn OUTPUT_FN]
                         [--quiet]

Create Optics Groups from EPU AFIS data

optional arguments:
  -h, --help            show this help message and exit
  --xml_dir XML_DIR, -i XML_DIR
                        Path to directory with EPU record metadata XML files.
                        [REQUIRED]
  --n_clusters N_CLUSTERS, -n N_CLUSTERS
                        Number of Optics Groups (1 = interactively choose. [1]
  --apix APIX           Pixel size of micrographs. [1.00]
  --mtf_fn MTF_FN       Path to MTF STAR file. [MTF.star]
  --voltage VOLTAGE     Voltage of micrscope in keV. [300.0]
  --cs CS               Spherical Aberration of objective lens in mm. [2.7]
  --q0 Q0               Fraction of amplitude contrast [0 - 1]. [0.1]
  --ftype FTYPE         Filetype of micrograph movies. [mrc]
  --movie_dir MOVIE_DIR, -d MOVIE_DIR
                        Path to directory with micrograph movies. [.]
  --output_fn OUTPUT_FN, -o OUTPUT_FN
                        Path to output STAR file. [movies.star]
  --quiet, -q           Run non-interactively with no plotting. [False]

Writted by Dustin Morado 20.06.2020
```

## Licensing

Distributed under the MIT License please see LICENSE.txt for more information.


CHANGELOG
v0.2
Changed to actually look for movie for when Scipion renames them. It runs a bit
slower though now looking for every movie.


EPU_Group_AFIS.py
#!/usr/bin/env python
#. -*- coding: utf-8 -*-
"""Sort EPU AFIS data into Optics Groups for RELION 3.1

Written By Dustin Reed Morado
Last updated 20.06.2020
"""

from __future__ import print_function, division

import os
import argparse
import numpy as np
import matplotlib.pyplot as plt

from fnmatch import fnmatch
from xml.etree import ElementTree
from sklearn.cluster import KMeans

def main(xml_dir = os.getcwd(), n_clusters = 1, apix = 1.00,
         mtf_fn = 'MTF.star', voltage = 300.0, cs = 2.7, q0 = 0.1,
         ftype = 'mrc', movie_dir = '', output_fn = 'movies.star',
         quiet = False):

    metadata_fns = []

    for dirpath, dirnames, filenames in os.walk(xml_dir):
        for filename in filenames:
            if (fnmatch(filename, 'FoilHole_*_Data_*.xml')
                    and not fnmatch(filename, '*_Fractions.xml')):

                metadata_fns.append(os.path.join(dirpath, filename))

    if len(metadata_fns) == 0:
        raise ValueError('No EPU XML metadata files found')

    namespace = {'fei' :
            'http://schemas.datacontract.org/2004/07/Fei.SharedObjects'}

    beam_shifts = []

    for metadata_fn in metadata_fns:
        metadata = ElementTree.parse(metadata_fn)
        beam_shift = metadata.findall(
                'fei:microscopeData/fei:optics/fei:BeamShift', namespace)

        if len(beam_shift) == 0:
            raise ValueError('No BeamShift found in {}'.format(metadata_fn))
        elif len(beam_shift[0]) != 2:
            raise ValueError('Improper BeamShift found in {}'.format(
                    metadata_fn))

        beam_shifts.append([float(x.text) for x in beam_shift[0]])

    shift_array = np.array(beam_shifts)

    if n_clusters == 1:
        figure = plt.figure()
        axes = figure.add_subplot(111)
        axes.set_title('EPU AFIS Beam Shift Values')
        axes.set_xlabel('Beam Shift X (a.u.)')
        axes.set_ylabel('Beam Shift Y (a.u.)')
        axes.scatter(shift_array[:, 0], shift_array[:, 1])
        plt.show(block = False)

        while n_clusters <= 1:
            n_user = input('How many Optics Groups (or Ctrl-C to abort): ')

            try:
                n_clusters = int(n_user)
            except ValueError:
                print('Please enter a positive integer greater than 1!')
            else:
                if n_clusters <=1:
                    print('Please enter a positive integer greater than 1!')

        plt.close('all')

    if len(beam_shifts) < n_clusters:
        raise ValueError('Number of clusters greater than number of points')

    k_means = KMeans(n_clusters = n_clusters)
    k_means.fit(shift_array)
    optics_clusters = k_means.predict(shift_array)
    cluster_centers = k_means.cluster_centers_

    radii = [x[0]**2 + x[1]**2 for x in cluster_centers]
    sort_idxs = sorted(enumerate(radii), key=lambda x:x[1])
    origin = cluster_centers[sort_idxs[0][0]]

    x_angle = np.pi
    y_angle = np.pi
    x_vec = (1, 0)
    y_vec = (0, 1)

    for idx in range(1,5):
        point = cluster_centers[sort_idxs[idx][0]]
        angle = np.arctan2(point[1] - origin[1], point[0] - origin[0])

        if abs(angle) < x_angle:
            x_angle = angle
            x_vec = (point[0] - origin[0], point[1] - origin[1])

        if abs(angle - (np.pi / 2)) < abs(y_angle - (np.pi / 2)):
            y_angle = angle
            y_vec = (point[0] - origin[0], point[1] - origin[1])

    grid_dists = []

    for idx, center in enumerate(cluster_centers):
        point = (center[0] - origin[0], center[1] - origin[1])
        grid_x = np.round(   (point[0] * x_vec[0] + point[1] * x_vec[1])
                           / (x_vec[0]**2 + x_vec[1]**2))

        grid_y = np.round(   (point[0] * y_vec[0] + point[1] * y_vec[1])
                           / (y_vec[0]**2 + y_vec[1]**2))

        dist = max(abs(grid_x), abs(grid_y))
        angle = (np.degrees(np.arctan2(grid_y, grid_x)) + 360) % 360
        grid_dists.append((idx, dist, angle))

    sort_idxs = [x[0] for x in sorted(grid_dists, key=lambda x:(x[1], x[2]))]
    optics_groups = [sort_idxs.index(x) + 1 for x in optics_clusters]

    if not quiet:
        figure = plt.figure()
        axes = figure.add_subplot(111)
        axes.set_title('EPU AFIS Beam Shift Values')
        axes.set_xlabel('Beam Shift X (a.u.)')
        axes.set_ylabel('Beam Shift Y (a.u.)')
        axes.scatter(shift_array[:, 0], shift_array[:, 1], c=optics_groups,
                cmap='tab20')

        for optics_group in range(n_clusters):
            axes.annotate('{0:d}'.format(optics_group + 1),
                    xy = cluster_centers[sort_idxs[optics_group]],
                    textcoords = 'offset pixels', xytext=(10, 10))

        plt.show(block = False)
        _ = input('If happy with cluster press any key (Ctrl-C to abort)')
        plt.close('all')

    entries = sorted(zip(metadata_fns, optics_groups),
            key=lambda x:(x[0], x[1]))

    with open(output_fn, 'w') as f:
        f.write('\n# version 30001\n\n')
        f.write('data_optics\n\nloop_\n')
        f.write('_rlnOpticsGroupName #1 \n')
        f.write('_rlnOpticsGroup #2 \n')
        f.write('_rlnMtfFileName #3 \n')
        f.write('_rlnMicrographOriginalPixelSize #4 \n')
        f.write('_rlnVoltage #5 \n')
        f.write('_rlnSphericalAberration #6 \n')
        f.write('_rlnAmplitudeContrast #7 \n')

        for optics_group in range(1, n_clusters + 1):
            f.write(('opticsGroup{0:d}\t{1:d}\t{2}\t{3:15.6f}\t{4:15.6f}\t'
                     '{5:15.6f}\t{6:15.6f}\n').format(optics_group,
                     optics_group, mtf_fn, apix, voltage, cs, q0))

        f.write(' \n\n# version 30001\n\n')
        f.write('data_movies\n\nloop_\n')
        f.write('_rlnMicrographMovieName #1 \n')
        f.write('_rlnOpticsGroup #2 \n')

        for metadata_fn, optics_group in entries:
            base = os.path.basename(metadata_fn)
            root, ext = os.path.splitext(base)
            movie_glob = '*{0}*_Fractions.{1}'.format(root, ftype)
            movie_fn = None

            for dirpath, dirnames, filenames in os.walk(movie_dir):
                for filename in filenames:
                    if fnmatch(filename, movie_glob):
                        movie_fn = os.path.join(dirpath, filename)
                        break

                if movie_fn is not None:
                    break

            if movie_fn is None:
                movie_fn = '{0}_Fractions.{1}'.format(
                        os.path.join(movie_dir, root), ftype)

            f.write('{0}\t{1:d}\n'.format(movie_fn, optics_group))

        f.write(' \n')

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
            description = "Create Optics Groups from EPU AFIS data",
            epilog = "Writted by Dustin Morado 20.06.2020")

    parser.add_argument('--xml_dir', '-i', type = str, required = True,
            help = ('Path to directory with EPU record metadata XML files. '
                    '[REQUIRED]'))

    parser.add_argument('--n_clusters', '-n', type = int, default = 1,
            help = 'Number of Optics Groups (1 = interactively choose. [1]')

    parser.add_argument('--apix', type = float, default = 1.00,
            help = 'Pixel size of micrographs. [1.00]')

    parser.add_argument('--mtf_fn', type = str, default = 'MTF.star',
            help = 'Path to MTF STAR file. [MTF.star]')

    parser.add_argument('--voltage', type = float, default = 300.0,
            help = 'Voltage of micrscope in keV. [300.0]')

    parser.add_argument('--cs', type = float, default = 2.7,
            help = 'Spherical Aberration of objective lens in mm. [2.7]')

    parser.add_argument('--q0', type = float, default = 0.1,
            help = 'Fraction of amplitude contrast [0 - 1]. [0.1]')

    parser.add_argument('--ftype', type = str, default = 'mrc',
            help = 'Filetype of micrograph movies. [mrc]')

    parser.add_argument('--movie_dir', '-d', type = str, default = '.',
            help = 'Path to directory with micrograph movies. [.]')

    parser.add_argument('--output_fn', '-o', type = str,
            default = 'movies.star',
            help = 'Path to output STAR file. [movies.star]')

    parser.add_argument('--quiet', '-q', action = 'store_true',
            help = 'Run non-interactively with no plotting. [False]')

    args = parser.parse_args()
    main(xml_dir = args.xml_dir,
         n_clusters = args.n_clusters,
         apix = args.apix,
         mtf_fn = args.mtf_fn,
         voltage = args.voltage,
         cs = args.cs,
         q0 = args.q0,
         ftype = args.ftype,
         movie_dir = args.movie_dir,
         output_fn = args.output_fn,
         quiet = args.quiet)

